### PR TITLE
Fix torch.compile produce wrong output due to problem in decomposition

### LINKED ISSUE #182012: `torch.compile` silently produces all-zero `x.grad` for `diagonal_scatter(...).sum().backward()`
## 🐛 Describe the bug

`torch.compile(backend='inductor')` silently produces `x.grad = 0` for
`torch.diagonal_scatter(x, src, offset).sum().backward()`, while eager mode
correctly produces `x.grad = 1` at all non-diagonal positions (and `0` on the
diagonal, since those positions were overwritten by `src`).

The forward output is correct; only the backward gradient for `x` is wrong.
`src.grad` is computed correctly. `aot_eager` also produces the correct gradient,
confirming the bug is in Inductor's code generation, not in AOTAutograd or the
decomposition.

## Minimal reproducer

```python
import torch

x = torch.randn(8, 8, device='cuda', requires_grad=True)
src = torch.randn(8, device='cuda', requires_grad=True)

def fn(x, src):
    return torch.diagonal_scatter(x, src, 0).sum()

# Eager: correct
x1 = x.clone().detach().requires_grad_(True)
s1 = src.clone().detach().requires_grad_(True)
fn(x1, s1).backward()
print(f"eager x.grad:\n{x1.grad}")
# tensor([[0., 1., 1., 1., 1., 1., 1., 1.],
#         [1., 0., 1., 1., 1., 1., 1., 1.],
#         ...])  # 1 everywhere except diagonal

# Compiled: wrong
torch._dynamo.reset()
x2 = x.clone().detach().requires_grad_(True)
s2 = src.clone().detach().requires_grad_(True)
torch.compile(fn)(x2, s2).backward()
print(f"compiled x.grad:\n{x2.grad}")
# tensor([[0., 0., 0., 0., 0., 0., 0., 0.],
#         [0., 0., 0., 0., 0., 0., 0., 0.],
#         ...])  # ALL ZEROS — incorrect
```

## Root cause analysis

The backward FX graph for `diagonal_scatter(x, src).sum()` is:

```
expand(tangent_scalar, [N, N])  →  clone  →  diagonal_scatter(clone, zeros)  →  return
```

Inductor lowers `clone(expand(scalar, [N, N]))` into a Triton kernel that
writes a single element to `buf0`, then returns
`reinterpret_tensor(buf0, (N, N), (0, 0), 0)` — a stride-(0, 0) view where
all N² positions alias one memory cell. The subsequent `diagonal_scatter`
kernel writes `0.0` to that single cell (to zero the diagonal), so the entire
gradient reads back as 0.

Correct behavior: `clone` must materialize the expanded tensor into a
contiguous buffer so that `diagonal_scatter` can selectively zero only the
diagonal positions.

This is **not** the same root cause as #180164 / #180771 (slice_scatter
backward), which was in AOTAutograd's `meta_slice_scatter` preserving zero
strides. Here `aot_eager` is correct; only Inductor's codegen is wrong.

## Affected configurations

- **Shapes**: all tested sizes (4×4 through 256×256), square and non-square, all diagonal offsets
- **Dtypes**: float32 and float64
- **Devices**: CUDA (tested on Tesla T4)
- **Trigger pattern**: `diagonal_scatter(x, src, offset)` followed directly by `.sum()` (any dim variant)
- Adding a pointwise op between `diagonal_scatter` and `sum` (e.g. `.relu()`, `.mul(2)`) avoids the bug by preventing the problematic fusion

## Characterization

| Pattern | x.grad |
|---------|--------|
| `diagonal_scatter(x, src, 0).sum().backward()` | **all zeros (BUG)** |
| `diagonal_scatter(x, src, 0).relu().sum().backward()` | correct |
| `diagonal_scatter(x, src, 0).mul(2).sum().backward()` | correct |
| `diagonal_scatter(x, src, 0).add(1).sum().backward()` | **all zeros (BUG)** |
| `diagonal_scatter(x, src, 0).clone().sum().backward()` | **all zeros (BUG)** |

## Versions

- **PyTorch**: 2.13.0.dev20260425+cu126
- **Triton**: 3.2.0+git4b3bb1e8
- **CUDA**: 12.6
- **GPU**: Tesla T4 (sm_75)
- **Python**: 3.11

cc @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by irshadcc**
Fixes #182012 

```python
import torch
def fn(x, src):
    return torch.diagonal_scatter(x, src, 0).sum()

x2 = x.clone().detach().requires_grad_(True)
s2 = src.clone().detach().requires_grad_(True)
torch.compile(fn)(x2, s2).backward()
print(f"compiled x2.grad = \n{x2.grad}")
```

Actual IR 
```text
op1: SchedulerNode(ComputedBuffer)
op1.writes = [MemoryDep('buf1', 0, {})]
op1.unmet_dependencies = [StarDep(name='buf0', mode=None)]
op1.met_dependencies = []
op1.outputs = [
    buf1: ComputedBuffer
    buf1.layout = MutationLayoutSHOULDREMOVE('cuda:0', torch.float32, size=[8], stride=[0, 0])
    buf1.mutations = ['buf0']
    buf1.users = [NodeUser(node=OUTPUT, can_inplace=False, is_weak=False)]
]
op1.group.device = cuda:0
op1.group.iteration = (8, 1)
op1.sizes = ([8], [])
buf0_layout = FixedLayout('cuda:0', torch.float32, size=[1, 1], stride=[0, 0])
buf1_layout = MutationLayoutSHOULDREMOVE('cuda:0', torch.float32, size=[8], stride=[0, 0])
class op1_loop_body:
    var_ranges = {p0: 8}
    index0 = 0
    def body(self, ops):
        constant = ops.constant(0.0, torch.float32)
        get_index = self.get_index('index0')
        store = ops.store('buf1', get_index, constant, None)
        return store
```

Expected IR 
```text
op1: SchedulerNode(ComputedBuffer)
op1.writes = [MemoryDep('buf1', 9*d0, {d0: 8})]
op1.unmet_dependencies = [StarDep(name='buf0', mode=None)]
op1.met_dependencies = []
op1.min_input_distance = 1
op1.max_input_distance = 1
op1.outputs = [
    buf1: ComputedBuffer
    buf1.layout = MutationLayoutSHOULDREMOVE('cuda:0', torch.float32, size=[8], stride=[8, 1])
    buf1.mutations = ['buf0']
    buf1.users = [NodeUser(node=OUTPUT, can_inplace=False, is_weak=False)]
]
op1.group.device = cuda:0
op1.group.iteration = (8, 1)
op1.sizes = ([8], [])
buf0_layout = FixedLayout('cuda:0', torch.float32, size=[8, 8], stride=[8, 1])
buf1_layout = MutationLayoutSHOULDREMOVE('cuda:0', torch.float32, size=[8], stride=[8, 1])
class op1_loop_body:
    var_ranges = {p0: 8}
    index0 = 9*p0
    def body(self, ops):
        constant = ops.constant(0.0, torch.float32)
        get_index = self.get_index('index0')
        store = ops.store('buf1', get_index, constant, None)
        return store
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
@claude explain those CI failures, check if they are related, and review the changes in this PR

**Review by jansel — CHANGES_REQUESTED** [review]
fix test failures

**Comment by irshadcc** [comment]
@claude please fix the test failures.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by irshadcc** [comment]
@jansel  Can we merge this ?

**Comment by jansel** [comment]
@pytorchbot drci

bot comment is out of date, I see "19 failing" below

**Comment by jansel** [comment]
@claude are those CI failures related?

**Review by jansel — APPROVED** [review]

**Comment by jansel** [comment]
@pytorchbot rebase

Let's see if this helps the tests.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by irshadcc** [comment]
@jansel The CI failures seems to be not related. I think the PR is ready for merge ?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by irshadcc** [comment]
@jansel Can we merge this PR ?

**Comment by jansel** [comment]
@claude explain those CI failures, are they caused by this PR?

**Comment by jansel** [comment]
@irshadcc I'd like to get CI green before merging.  Can you rebase these changes to viable/strict or main.

**Comment by irshadcc** [comment]
Till now, I thought that the failure was unrelated as claimed by claude. Today I looked into the issue and it looks like it was indeed related. [VariableBuilder::wrap_numpy_ndarray](https://github.com/pytorch/pytorch/blob/d34af8fdfbb80527765f61e669c5584a50614686/torch/_dynamo/variables/builder.py#L3089) uses clone_preserve_strides and my previous changes changed the strides and lead to guard failure. A safer fix is to move my changes to the decomposition function.

**Comment by irshadcc** [comment]
It seems some tests are failing, I will fix it and push once I make sure all of them succeed

**Comment by irshadcc** [comment]
@claude How to run the CI and check if all the CI is passing ?

**Comment by jansel** [comment]
@claude can you confirm that the new semantics of the decomp match what is in eager mode
